Add ETW trace for PerformDependencyAnalysis

Forgind opened the pull request:
I was considering adding bucket.bucketSequenceNumber, but I don't think that's too important. Complicated information can't be easily ( / efficiently) attached to ETW traces

Fixes #6616

I should make sure they appear in PerfView before merging.

rainersigwald commented on src/Build/BackEnd/Components/RequestBuilder/TargetEntry.cs:
```diff
@@ -465,7 +466,9 @@ internal async Task ExecuteTarget(ITaskBuilder taskBuilder, BuildRequestEntry re
 
                         // UNDONE: (Refactor) Refactor TargetUpToDateChecker to take a logging context, not a logging service.
                         TargetUpToDateChecker dependencyAnalyzer = new TargetUpToDateChecker(requestEntry.RequestConfiguration.Project, _target, targetLoggingContext.LoggingService, targetLoggingContext.BuildEventContext);
+                        MSBuildEventSource.Log.PerformDependencyAnalysisStart();
```
I'd claim that creating the object is part of the time spent (I doubt this matters much though).

```suggestion
                        MSBuildEventSource.Log.PerformDependencyAnalysisStart();
                        TargetUpToDateChecker dependencyAnalyzer = new TargetUpToDateChecker(requestEntry.RequestConfiguration.Project, _target, targetLoggingContext.LoggingService, targetLoggingContext.BuildEventContext);
```

rainersigwald commented on src/Build/BackEnd/Components/RequestBuilder/TargetEntry.cs:
```diff
@@ -465,7 +466,9 @@ internal async Task ExecuteTarget(ITaskBuilder taskBuilder, BuildRequestEntry re
                         TargetUpToDateChecker dependencyAnalyzer = new TargetUpToDateChecker(requestEntry.RequestConfiguration.Project, _target, targetLoggingContext.LoggingService, targetLoggingContext.BuildEventContext);
+                        MSBuildEventSource.Log.PerformDependencyAnalysisStart();
                         DependencyAnalysisResult dependencyResult = dependencyAnalyzer.PerformDependencyAnalysis(bucket, out changedTargetInputs, out upToDateTargetInputs);
+                        MSBuildEventSource.Log.PerformDependencyAnalysisStop();
```
Can this log the result? I think we'll be interested in two aspects:

1. Time it takes to decide something's fully up to date (pure overhead in a no-op incremental build)
2. Time it takes to decide something's _not_ up to date in "obvious" cases (overhead when we need to build).

Forgind replied:
It definitely can...I don't think it can log what the result really means, though. In other words, it can log the int corresponding to the enum, but I don't think it can log what "0" or "4" correspond to. I can add it in anyway, since you can always look it up if it's important to you.

rainersigwald replied:
That's good enough for me too.

rainersigwald commented on src/Framework/MSBuildEventSource.cs:
```diff
@@ -423,6 +423,18 @@ public void PacketReadSize(int size)
         }
+
+        [Event(56, Keywords = Keywords.All)]
+        public void PerformDependencyAnalysisStart()
```
I think I'd prefer a name like `TargetUpToDateCheckStart`/Stop; as the comment above notes this "isn't really dependency analysis" so I'd love to avoid propagating the bad name. Thoughts?

Forgind replied:
I slightly prefer it this way because 1) there is other work that happens when it does an UTD check, albeit not too much, and 2) consistency is good even if it's slightly wrong consistency. I don't feel at all strongly, though, so I can change it.

rainersigwald reviewed (changes requested):
This is probably the most interesting part of #6616 but not all of it -- we'll need some work in Copy and other unconditionally-called tasks to call that done.

Forgind commented:
I added an ETW event to Copy. I'd looked it over before, but it mostly seemed captured by the ExecuteTask ETW. I added something that notes whether it's parallel, but I'm now thinking it would have made more sense to specify how parallel, i.e., an int for parallelism (if parallel) or 1 if not. It also might be useful to put it under a IsEnabled check and include destinationFilesSuccessfullyCopied.Select(f => f.ItemSpec) on Stop events. Will do if you think it's a good idea.

I'm not really sure how to find other tasks that run unconditionally, since you suggested away from debugging, and I couldn't (quickly) find documentation on it.

rainersigwald commented on src/Framework/MSBuildEventSource.cs:
```diff
@@ -423,6 +423,30 @@ public void PacketReadSize(int size)
+        }
+
+        [Event(58, Keywords = Keywords.All)]
+        public void CopyStart(bool singleThreaded)
```
I think CopyStart/Stop is well covered by ExecuteTaskStart/Stop events. I was hoping for events measuring the up-to-date check within Copy.

rainersigwald commented:
> I'm not really sure how to find other tasks that run unconditionally

[Targets run incrementally](https://docs.microsoft.com/visualstudio/msbuild/incremental-builds) if they define inputs and outputs and the outputs are up to date with respect to the inputs' timestamps.

Some targets do not define inputs and outputs. This is sometimes a bug (especially in custom targets/additions to the standard process) but sometimes intentional behavior because incremental behavior happens at another layer (like inside the Copy task or with tracker-based incremental builds for C++ compilation), can't be represented (like in the case of RAR), or isn't important (targets that just mutate MSBuild item/property state and are quick).

One way to do this would be to create a template project, build it, build it again, and see what tasks ran the second time through. Another would be to explicitly look for targets that don't have inputs/outputs.
